src/telemetry.ts

```diff
@@ -0,0 +1,38 @@
+import TelemetryReporter from 'vscode-extension-telemetry';
+import { BaseLanguageClient, ClientCapabilities, StaticFeature } from 'vscode-languageclient';
+
+const TELEMETRY_VERSION = 1;
+
+type TelemetryEvent = {
+  v: number;
+  name: string;
+  properties: { string };
+};
+
+export class TelemetryFeature implements StaticFeature {
+  constructor(private client: BaseLanguageClient, private reporter: TelemetryReporter) {
+    this.client.onTelemetry((event: TelemetryEvent) => {
```

Comment:
`onTelemetry` returns a Disposable. Do we need to keep track of it and dispose it in `this.dispose`? Maybe @jpogran knows what's the convention for this.

Reply:
I don't think so, as they are just events. If you were to keep track to call in dispose, you'd be keeping track of possibly hundreds or thousands of events to dispose at extension deactivate. That seems like a memory problem there.

What I'm more concerned about is that the `StaticFeature` class says `initialize` is called after the client receives the initialize request from the server but before the client sends the initialized notification to the server. We're setting up the handler in the constructor which isn't part of the above guarantee. According to https://github.com/Microsoft/vscode-languageserver-node#303-client-server-and-json-rpc, you must use the `send|on` methods after `onReady` returns, otherwise exceptions can happen. I think we should move this to the `initialize` method to ensure it's called at the correct time.

With that said, I'm not sure this needs to be a `StaticFeature`. This all occurs only if the LS sends a telemetry event. Registering a handler for something that never happens isn't an error condition. I think we could make this a three line addition to `ClientHandler` instead of an entire new class and initialization concern. Thoguhts?

Reply:
That is a good point @jpogran - I will update the PR!

Reply:
@jpogran actually I don't see how else I could fill in the client capabilities without using either `StaticFeature` or `DynamicFeature` https://github.com/microsoft/vscode-languageserver-node/blob/f16ca298dac93a99b2338129a2149a0ada9021dc/client/src/common/client.ts#L3716-L3760

Reply:
> I think we should move this to the initialize method to ensure it's called at the correct time.
> https://github.com/Microsoft/vscode-languageserver-node#303-client-server-and-json-rpc

Strictly speaking we do not have any telemetry events coming from the server before `initialize`, so this probably won't matter today, but what you linked above seems to contradict LSP spec, which does allow telemetry notifications to be sent before:
https://microsoft.github.io/language-server-protocol/specifications/specification-current/#initialize

> Until the server has responded to the `initialize` request with an `InitializeResult`, the client must not send any additional requests or notifications to the server. In addition the server is not allowed to send any requests or notifications to the client until it has responded with an `InitializeResult`, with the exception that during the `initialize` request the server is allowed to send the notifications `window/showMessage`, `window/logMessage` and **`telemetry/event`** as well as the `window/showMessageRequest` request to the client.

Reply:
Here I'm assuming that `onReady()` would only return when initialization is finished, i.e. before `initialized` is sent from the client - please do correct me if I'm wrong.